### PR TITLE
Refactor Managed API to accept and return individual messages

### PR DESCRIPTION
**Pull request by kkohbrok**
This PR refactors the functions in the Managed API to accept individual messages (`process_message`) and return individual messages (`propose_...`, etc).

This necessitates some refactoring of the events returned and in particular allows us to return errors if messages are invalid as opposed to just putting error events into the queue.

### CODE REVIEW

File: src/group/managed_group/mod.rs [hunk 1]
```diff
@@ -802,67 +755,60 @@ impl ManagedGroup {
...
+                    true, /* force_self_update */
+                    None,
+                )?
             }
         };
+
+        // Take the new KeyPackageBundle and save it for later
+        let kpb = kpb_option.ok_or(ManagedGroupError::LibraryError(
+            "We didn't get a key package for a full commit on self update.".into(),
+        ))?;
```
Comment (franziskuskiefer) [inline src/group/managed_group/mod.rs:792]: What clippy says.

File: src/group/managed_group/mod.rs [hunk 2]
```diff
@@ -956,24 +902,20 @@ impl ManagedGroup {
...
-                }
-            };
-            out.push(msg);
-        }
-        Ok(out)
+        plaintext: MlsPlaintext,
+    ) -> Result<MlsMessage, ManagedGroupError> {
+        let msg = match self.configuration().handshake_message_format {
+            HandshakeMessageFormat::Plaintext => MlsMessage::Plaintext(plaintext),
+            HandshakeMessageFormat::Ciphertext => {
```
Comment (franziskuskiefer) [inline src/group/managed_group/mod.rs:911]: Codecov suggests that we never use this with `Ciphertext`. Can you check if this is really the case?

File: tests/utils/managed_utils/client.rs [hunk 3]
```diff
@@ -175,12 +173,16 @@ impl Client {
         let action_results = match action_type {
             ActionType::Commit => {
                 let (messages, welcome) = group.add_members(&self.key_store, key_packages)?;
-                (messages, Some(welcome))
+                (vec![messages], Some(welcome))
+            }
+            ActionType::Proposal => {
+                let mut messages = Vec::new();
+                for key_package in key_packages {
```
Comment (franziskuskiefer) [inline tests/utils/managed_utils/client.rs:180]: This looks like it should be a `for_each()` with `collect()`.

File: tests/utils/managed_utils/client.rs [hunk 4]
```diff
@@ -201,11 +203,19 @@ impl Client {
...
-                None,
-            ),
+            ActionType::Commit => {
+                let (message, welcome_option) =
+                    group.remove_members(&self.key_store, target_indices)?;
+                (vec![message], welcome_option)
+            }
+            ActionType::Proposal => {
+                let mut messages = Vec::new();
+                for &target_index in target_indices {
```
Comment (franziskuskiefer) [inline tests/utils/managed_utils/client.rs:213]: Same here.

File: src/group/managed_group/mod.rs [hunk 5]
```diff
@@ -802,67 +755,60 @@ impl ManagedGroup {
...
-            &credential_bundle,
-            &messages_to_commit,
-            &[],
-            true, /* force_self_update */
-            None,
-        )?;
-
-        // Add the Commit message to the other pending messages
-        plaintext_messages.push(commit);
+        let messages_to_commit: Vec<&MlsPlaintext> = self.pending_proposals.iter().collect();
```
Comment (franziskuskiefer) [inline src/group/managed_group/mod.rs:759]: Having to collect something here sounds like the APIs are bad. But that's probably nothing to solve in this PR.

File: src/group/managed_group/mod.rs [hunk 6]
```diff
@@ -437,170 +418,145 @@ impl ManagedGroup {
...
-                    }
+        // Check the type of message we received
+        let (plaintext, aad_option) = match message {
+            // If it is a ciphertext we decrypt it and return the plaintext message
+            MlsMessage::Ciphertext(ciphertext) => {
+                let aad = ciphertext.authenticated_data.clone();
+                (
+                    self.group
+                        .decrypt(&ciphertext)
+                        .map_err(|e| InvalidMessageError::InvalidCiphertext(e))?,
```
Comment (franziskuskiefer) [inline src/group/managed_group/mod.rs:437]: What clippy says.

File: src/group/managed_group/mod.rs [hunk 7]
```diff
@@ -437,170 +418,145 @@ impl ManagedGroup {
...
-                            events.push(GroupEvent::InvalidMessage(InvalidMessageEvent::new(
-                                InvalidMessageError::InvalidCiphertext(aad.into()),
-                            )));
-                            // Since we cannot decrypt the MlsCiphertext to a MlsPlaintext we move
-                            // to the next message
-                            continue;
-                        }
-                    }
+        // Check the type of message we received
+        let (plaintext, aad_option) = match message {
```
Comment (franziskuskiefer) [inline src/group/managed_group/mod.rs:430]: This entire function needs to be refactored... maybe file a follow up.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Generally looks good to me. Mostly nits and some follow ups.

File: tests/utils/managed_utils/client.rs (hunk 3 above)
Comment (kkohbrok) [reply tests/utils/managed_utils/client.rs:180]: Good idea! I'm not sure how I should handle the potential error thrown by `propose_add_member` in the `for_each` closure, though. Same in the other code below.

**Review by kkohbrok — COMMENTED** [review]
Thanks for the review! I've addressed what I could and otherwise replied to the comments.

File: src/group/managed_group/mod.rs (hunk 2 above)
Comment (franziskuskiefer) [reply src/group/managed_group/mod.rs:913]: Do you want to add a test (or parameterise some tests to use ciphertexts)? If not, please file a follow up.

File: src/group/managed_group/errors.rs [hunk 8]
```diff
@@ -36,6 +37,7 @@ implement_error! {
             EmptyInput(EmptyInputError) =
                 "Empty input. Additional detail is provided.",
             KeyStoreError(KeyStoreError) = "See [`KeyStoreError`](`crate::key_store::KeyStoreError`) for details",
+            InvalidMessage(InvalidMessageError) = "The message could not be processed.",
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/group/managed_group/errors.rs:40]: Can you file a follow up then for a negative test case?

File: tests/utils/managed_utils/client.rs (hunk 3 above)
Comment (franziskuskiefer) [reply tests/utils/managed_utils/client.rs:180]: Ah, I didn't see the error there. Then let's leave it like this for now.

**Review by franziskuskiefer — APPROVED** [review]
lgtm, with the remaining comment addressed in some way.

File: src/group/managed_group/errors.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kkohbrok) [reply src/group/managed_group/errors.rs:40]: I've included a link to this conversation in #133.

File: src/group/managed_group/mod.rs (hunk 2 above)
Comment (kkohbrok) [reply src/group/managed_group/mod.rs:916]: I've changed one of the managed group tests to use ciphertexts. That should increase the coverage.

**Review by franziskuskiefer — APPROVED** [review]
:shipit:

**Pull request merged**